File: src/bip322.ts
import { init, g, unwrap } from './init';
import type { Bytes, VerifyResult, Network } from './types';

/** BIP-322 generic signed message verification and signing helpers. */
export const bip322 = {
  /** Verify a BIP-322 signed message.
   *  Calls Go: bip322.VerifyMessage() from btcutil/bip322
   *  (after btcutil.DecodeAddress() to parse the address).
   *
   *  @param message   - The message that was signed.
   *  @param address   - The Bitcoin address to verify against.
   *  @param signature - The base64-encoded BIP-322 signature.
   *  @param network   - Bitcoin network (default: "mainnet").
   *  @returns An object with `valid` (boolean) and optionally `error` (string). */
  async verifyMessage(
    message: string,
    address: string,
    signature: string,
    network: Network = 'mainnet',
  ): Promise<VerifyResult> {
    await init();
    return unwrap<VerifyResult>(
      g().bip322.verifyMessage(
        message,
        address,
        signature,
        network,
      )
    );
  },

  /** Build the BIP-322 "to_sign" PSBT for the simple format.
   *  Calls Go: bip322.BuildToSignPacketSimple() from btcutil/bip322.
   *
   *  @param message  - The message to sign.
   *  @param pkScript - The pkScript of the signing address.
   *  @returns Base64-encoded PSBT ready for signing. */
  async buildToSignPacketSimple(
    message: string,
    pkScript: Bytes,
  ): Promise<string> {
    await init();
    return unwrap<string>(
      g().bip322.buildToSignPacketSimple(message, pkScript),
    );
  },

  /** Build the BIP-322 "to_sign" PSBT for the full format.
   *  Calls Go: bip322.BuildToSignPacketFull() from btcutil/bip322.
   *
   *  @param message      - The message to sign.
   *  @param pkScript     - The pkScript of the signing address.
   *  @param txVersion    - Transaction version (typically 2).
   *  @param lockTime     - Lock time (typically 0).
   *  @param sequence     - Input sequence (typically 0).
   *  @returns Base64-encoded PSBT ready for signing. */
  async buildToSignPacketFull(
    message: string,
    pkScript: Bytes,
    txVersion: number,
    lockTime: number,
    sequence: number,
  ): Promise<string> {
    await init();
    return unwrap<string>(
      g().bip322.buildToSignPacketFull(
        message, pkScript, txVersion, lockTime, sequence,
      ),
    );
  },

  /** Serialize a witness stack into the BIP-322 wire format.
   *  Calls Go: bip322.SerializeTxWitness() from btcutil/bip322.
   *
   *  @param witness - Array of witness items (hex strings or Uint8Arrays).
   *  @returns Serialized witness bytes (Uint8Array). */
  async serializeTxWitness(witness: Bytes[]): Promise<Uint8Array> {
    await init();
    return unwrap<Uint8Array>(
      g().bip322.serializeTxWitness(witness),
    );
  },

  /** Parse serialized witness bytes back into a witness stack.
   *  Calls Go: bip322.ParseTxWitness() from btcutil/bip322.
   *
   *  @param rawWitness - Serialized witness data (hex string or Uint8Array).
   *  @returns Array of witness stack items (Uint8Array[]). */
  async parseTxWitness(rawWitness: Bytes): Promise<Uint8Array[]> {
    await init();
    return unwrap<Uint8Array[]>(
      g().bip322.parseTxWitness(rawWitness),
    );
  },
};

File: wasm-wrapper/bip322.go
//go:build js && wasm

package main

import (
	"encoding/hex"
	"syscall/js"

	"github.com/btcsuite/btcd/btcutil"
	"github.com/btcsuite/btcd/btcutil/bip322"
	"github.com/btcsuite/btcd/wire"
)

func bip322VerifyMessage(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 3, "message, address, signature"); e != nil {
		return e
	}

	message := args[0].String()
	addrStr := args[1].String()
	sig := args[2].String()
	network := optString(args, 3, "mainnet")

	params, ok := networkParams[network]
	if !ok {
		return errfResult("unknown network: %s", network)
	}

	addr, err := btcutil.DecodeAddress(addrStr, params)
	if err != nil {
		return errfResult("decode address: %s", err)
	}

	valid, err := bip322.VerifyMessage(message, addr, sig)
	result := map[string]any{"valid": valid}
	if err != nil {
		result["error"] = err.Error()
	}
	return okResult(result)
}

// bip322BuildToSignPacketSimple wraps bip322.BuildToSignPacketSimple.
// Args: message (string), pkScript (Bytes)
// Returns: base64 PSBT string
func bip322BuildToSignPacketSimple(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 2, "message, pkScript"); e != nil {
		return e
	}
	pkScript, e := bytesFromArg(args[1])
	if e != nil {
		return e
	}
	pkt, err := bip322.BuildToSignPacketSimple(
		[]byte(args[0].String()), pkScript,
	)
	if err != nil {
		return errfResult("buildToSignPacketSimple: %s", err)
	}
	b64, err := pkt.B64Encode()
	if err != nil {
		return errfResult("encode: %s", err)
	}
	return okResult(b64)
}

// bip322BuildToSignPacketFull wraps bip322.BuildToSignPacketFull.
// Args: message (string), pkScript (Bytes), txVersion (int), lockTime (int), sequence (int)
// Returns: base64 PSBT string
func bip322BuildToSignPacketFull(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 5,
		"message, pkScript, txVersion, lockTime, sequence"); e != nil {
		return e
	}
	pkScript, e := bytesFromArg(args[1])
	if e != nil {
		return e
	}
	pkt := bip322.BuildToSignPacketFull(
		[]byte(args[0].String()), pkScript,
		int32(args[2].Int()),
		uint32(args[3].Int()),
		uint32(args[4].Int()),
	)
	b64, err := pkt.B64Encode()
	if err != nil {
		return errfResult("encode: %s", err)
	}
	return okResult(b64)
}

// bip322SerializeTxWitness wraps bip322.SerializeTxWitness.
// Args: witness (array of Bytes)
// Returns: Uint8Array
func bip322SerializeTxWitness(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "witness[]"); e != nil {
		return e
	}
	n := args[0].Length()
	witness := make(wire.TxWitness, n)
	for i := 0; i < n; i++ {
		item := args[0].Index(i)
		if item.InstanceOf(js.Global().Get("Uint8Array")) {
			buf := make([]byte, item.Length())
			if len(buf) > 0 {
				js.CopyBytesToGo(buf, item)
			}
			witness[i] = buf
		} else {
			b, err := hex.DecodeString(item.String())
			if err != nil {
				return errfResult("invalid witness[%d]: %s",
					i, err)
			}
			witness[i] = b
		}
	}
	serialized, err := bip322.SerializeTxWitness(witness)
	if err != nil {
		return errfResult("serializeTxWitness: %s", err)
	}
	return okResult(bytesToJS(serialized))
}

// bip322ParseTxWitness wraps bip322.ParseTxWitness.
// Args: rawWitness (Bytes)
// Returns: array of Uint8Array (witness stack items)
func bip322ParseTxWitness(_ js.Value, args []js.Value) any {
	if e := checkArgs(args, 1, "rawWitness"); e != nil {
		return e
	}
	raw, e := bytesFromArg(args[0])
	if e != nil {
		return e
	}
	witness, err := bip322.ParseTxWitness(raw)
	if err != nil {
		return errfResult("parseTxWitness: %s", err)
	}
	items := make([]any, len(witness))
	for i, w := range witness {
		items[i] = bytesToJS(w)
	}
	return okResult(items)
}
